### PR TITLE
Add wiki article page

### PR DESCRIPTION
**Pull request by gagahpangeran**
Also add sidebar in this changes.

Screenshot from 2021-06-04 18-11-07

### CODE REVIEW

File: osu.Game/Overlays/Wiki/WikiSidebar.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                Padding = new MarginPadding { Left = subtitle ? 10 : 0 };
+            }
+
+            protected override IEnumerable<Drawable> EffectTargets => new Drawable[] { textFlow };
+
+            [BackgroundDependencyLoader]
+            private void load(OverlayColourProvider colourProvider)
+            {
+                IdleColour = colourProvider.Light2;
+                HoverColour = colourProvider.Light1;
```
Comment (bdach) [inline osu.Game/Overlays/Wiki/WikiSidebar.cs:106]: This hover colour spec actually does nothing, because the headings don't have an associated `Action` with them, which means they have `Enabled = false`. The stuff from `OnClick()` should be moved to `Action`.

Additionally, the colour should also remain `Light1` after click, as per web.

2021-06-04-142502_346x47_scrot

File: osu.Game/Overlays/Wiki/WikiSidebar.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+                    case LinkInline { IsImage: false } linkInline:
+                        return getTitle(linkInline);
+                }
+            }
+
+            return string.Empty;
+        }
+
+        private class TocTitle : OsuHoverContainer
```
Comment (bdach) [inline osu.Game/Overlays/Wiki/WikiSidebar.cs:72]: Please rename this to not use the `Toc` abbreviation. Full `TableOfContents` is fine.

Would also name this `TableOfContentsEntry` as it seems a more fitting name.

File: osu.Game/Overlays/Wiki/WikiSidebar.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        private string getTitle(ContainerInline containerInline)
+        {
+            foreach (var inline in containerInline)
+            {
+                switch (inline)
+                {
+                    case LiteralInline literalInline:
+                        return literalInline.Content.ToString();
+
+                    case LinkInline { IsImage: false } linkInline:
```
Comment (bdach) [inline osu.Game/Overlays/Wiki/WikiSidebar.cs:64]: We don't really use this syntax much anywhere else, it's a bit hard to parse visually here. A plain `if` will be fine.

File: osu.Game/Overlays/WikiOverlay.cs [hunk 4]
```diff
@@ -82,6 +89,17 @@ protected void LoadDisplay(Drawable display)
...
         }
 
+        protected override void UpdateAfterChildren()
+        {
+            base.UpdateAfterChildren();
+
+            if (articlePage != null)
+            {
+                articlePage.SidebarContainer.Height = DrawHeight;
+                articlePage.SidebarContainer.Y = Math.Clamp(ScrollFlow.Current - Header.DrawHeight, 0, Math.Max(ScrollFlow.ScrollContent.DrawHeight - DrawHeight - Header.DrawHeight, 0));
```
Comment (bdach) [inline osu.Game/Overlays/WikiOverlay.cs:99]: What's this about? I can't say I'm following what this is trying to do very well.

File: osu.Game/Overlays/Wiki/WikiArticlePage.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+using System;
+using Markdig.Syntax;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Containers.Markdown;
+using osu.Game.Overlays.Wiki.Markdown;
+
+namespace osu.Game.Overlays.Wiki
+{
+    public class WikiArticlePage : GridContainer
```
Comment (bdach) [inline osu.Game/Overlays/Wiki/WikiArticlePage.cs:13]: Should probably be a `CompositeDrawable` with the `GridContainer` as an `InternalChild`. Otherwise this is outwardly exposing `GridContainer` members, such as cell content setters, that shouldn't really be exposed.

File: osu.Game/Overlays/Wiki/WikiSidebar.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                Child = textFlow = new OsuTextFlowContainer(t =>
+                {
+                    t.Font = OsuFont.GetFont(size: subtitle ? 12 : 15);
+                }).With(f =>
+                {
+                    f.AddText(text);
+                    f.RelativeSizeAxes = Axes.X;
+                    f.AutoSizeAxes = Axes.Y;
+                });
+                Margin = new MarginPadding { Top = subtitle ? 5 : 10 };
```
Comment (bdach) [inline osu.Game/Overlays/Wiki/WikiSidebar.cs:96]: Don't think this is entirely right. As far as I can see from web, the spacing is:

* 2px bottom margin on `.wiki-toc-list__link`
* 5px bottom margin on `.wiki-toc-list--top > .wiki-toc-list__item`
* 10px bottom margin on `.wiki-toc-list`

In other words:

* All links (both top-level and sub-links) have a 2px bottom margin
* Singular top-level links without child entries have a 5px bottom margin
* Top-level links with child entries have a 10px bottom margin

File: osu.Game/Overlays/WikiOverlay.cs [hunk 7]
```diff
@@ -115,20 +133,7 @@ private void onSuccess(APIWikiPage response)
...
-                    Text = response.Markdown,
-                    DocumentMargin = new MarginPadding(0),
-                    DocumentPadding = new MarginPadding
-                    {
-                        Vertical = 20,
-                        Left = 30,
-                        Right = 50,
-                    },
-                });
+                LoadDisplay(articlePage = new WikiArticlePage($@"{api.WebsiteRootUrl}/wiki/{path.Value}/", response.Markdown));
```
Comment (bdach) [inline osu.Game/Overlays/WikiOverlay.cs:136]: Could also use `WikiArticlePage` in the `onFail()` scenario to decrease layout code duplication, I think?

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Overlays/Wiki/WikiSidebar.cs (hunk 1 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs:106]: Ah, I missed that.

File: osu.Game/Overlays/WikiOverlay.cs (hunk 7 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/WikiOverlay.cs:136]: Yup, it could do.

File: osu.Game/Overlays/Wiki/WikiSidebar.cs (hunk 3 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs:64]: When I code this, I use `case LinkInline linkInline when !linkInline.IsImage`. But then rider suggests that change and I just applied it.

There are many use of syntax `case ... when ...` in osu! and framework side. Should I just use `case ... when ...`?

File: osu.Game/Overlays/WikiOverlay.cs (hunk 4 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/WikiOverlay.cs:99]: Hmm, honestly I just copied this calculation from `NewsOverlay`.

https://github.com/ppy/osu/blob/10acad6524d94335f710b5619fdee0bf39c5b64a/osu.Game/Overlays/NewsOverlay.cs#L133-L138

File: osu.Game/Overlays/Wiki/WikiSidebar.cs (hunk 1 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs:106]: For now I can change the colour to `Light1` after click inside the `OnClick`. But I dont know how to change the colour back to `Light2` when the other entry is clicked.

Comment (bdach) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs:106]: You'd probably have to somehow track the last item clicked at the level of the table of contents sidebar. Maybe not especially worth pursuing right now.

File: osu.Game/Overlays/Wiki/WikiSidebar.cs (hunk 3 above)
Comment (bdach) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs:64]: I'm fine with `case...when`, but I do recall others had an issue with that syntax and the most uncontroversial choice was just using a plain `if`.

Comment (peppy) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs:64]: Please don't use case-when here.

File: osu.Game/Overlays/WikiOverlay.cs (hunk 4 above)
Comment (bdach) [reply osu.Game/Overlays/WikiOverlay.cs:98]: I see. Seems that this is used to implement a sticky container. It would be nice to not have this logic duplicated in two containers, though.

File: osu.Game/Overlays/Wiki/WikiSidebar.cs (hunk 1 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs]: Now for simplicity, I just remove the onclick and the colour change.

File: osu.Game/Overlays/Wiki/WikiSidebar.cs (hunk 6 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/Wiki/WikiSidebar.cs]: I refactor this and create `WikiTableOfContents`.

File: osu.Game/Overlays/WikiOverlay.cs (hunk 4 above)
Comment (gagahpangeran) [reply osu.Game/Overlays/WikiOverlay.cs:98]: Yeah, something like abstract class for both overlays?

Comment (bdach) [reply osu.Game/Overlays/WikiOverlay.cs:98]: Sounds like it could work, yes.

Comment (gagahpangeran) [reply osu.Game/Overlays/WikiOverlay.cs:98]: I think that's out of scope in this PR for now.

Comment (frenzibyte) [reply osu.Game/Overlays/WikiOverlay.cs:98]: Not really, should be a pretty simple change. I would expose a protected virtual property or something to get the `SidebarContainer` if that's what you're worried about.

Comment (bdach) [reply osu.Game/Overlays/WikiOverlay.cs:98]: I'm fine to proceed without it and do it as a follow-up, fwiw.

Comment (gagahpangeran) [reply osu.Game/Overlays/WikiOverlay.cs:98]: Yeah, I don't want to refactor the whole overlay in this PR.

File: osu.Game/Overlays/WikiOverlay.cs [hunk 8]
```diff
@@ -25,15 +25,21 @@ public class WikiOverlay : OnlineOverlay<WikiHeader>
...
         private CancellationTokenSource cancellationToken;
 
         private bool displayUpdateRequired = true;
 
+        private WikiArticlePage articlePage;
+
         public WikiOverlay()
             : base(OverlayColourScheme.Orange, false)
         {
+            scrollContainer = ScrollFlow;
```
Comment (bdach) [inline osu.Game/Overlays/WikiOverlay.cs:42]: This seems a little dangerous as it relies on the base initialising everything in ctor and not in BDL, for example. Think I'll just remove this field and hoist the `[Cached]` spec up to the original place of definition. Seems like a generally useful thing to cache anyways.

**Review by bdach — APPROVED** [review]
Have applied two small clean-ups, but generally looks good. :+1:
